### DISCUSSION TITLE
Hardware acceleration

### DISCUSSION THREAD

**Discussion by mattjohnsonpint** [ReactiveCircus/android-emulator-runner#286] [General]
This is a bit of continuation from #279.  (@mrk-han especially)

I've been doing some experimenting, and I don't think thatinstalling HAXM on MacOS under GitHub Actions hosted runners is actually doing anything whatsoever for the Android emulator.

`brew install --cask intel-haxm` does indeed say it succeeds, but it doesn't actually work.  Same with `sudo $ANDROID_HOME/cmdline-tools/latest/bin/sdkmanager --install "extras;intel;Hardware_Accelerated_Execution_Manager"` which does the same thing without brew.

- The [HAXM docs](https://github.com/intel/haxm/wiki/Installation-Instructions-on-macOS) say that you should see the HAXM kernel extension running with `kextstat | grep intel`, but that returns nothing.  Run without grep shows other extensions, but not HAXM.

- The [android docs here](https://developer.android.com/studio/run/emulator-acceleration#vm-mac) say that on macOS 10.10 and higher, the built-in `Hypervisor.Framework` is used first, and HAXM is a fallback.  It also says [here](https://developer.android.com/studio/known-issues#emulator_high_sierra) that on macOS 10.13 and higher one has to enable security permissions manually in the system preferences UI in order to load a third-party (non-apple) kernel extension at all.

This aligns with my findings.   After installing HAXM (with either brew or sdkmanager), `/Library/Extensions/intelhaxm.kext` is on disk, but not loaded.  `sudo kextload /Library/Extensions/intelhaxm.kext` gives an error:

```
Executing: /usr/bin/kmutil load -p /Library/Extensions/intelhaxm.kext
Error Domain=KMErrorDomain Code=27 "Extension with identifiers com.apple.nke.rvi,com.intel.kext.intelhaxm not approved to load. Please approve using System Preferences."
UserInfo={NSLocalizedDescription=Extension with identifiers com.apple.nke.rvi,com.intel.kext.intelhaxm not approved to load. Please approve using System Preferences.}
Error: Process completed with exit code 27.
```

And regardless of whether I install HAXM or not, checking the emulator with `$ANDROID_HOME/emulator/emulator -accel-check` always gives the same result:

```
accel:
0
Hypervisor.Framework OS X Version 12.
accel
```

I interpret this to mean that the Android emulator is indeed using hardware acceleration, as provided by Hypervisor.Framework, and I cannot *use* HAXM because I don't have ability to set the security permission programatically.

Am I wrong?

(I had the same results on `macos-11` or `macos-12` GitHub Actions hosted runners)

CC: @vaind (any thoughts?)

**Comment by vaind** [comment]
While I haven't done such a thorough investigation I've had my suspicion this is the case, too. 

Also, I have a [screen recording from CI](https://github.com/ReactiveCircus/android-emulator-runner/files/9908868/out.zip) that asks user to confirm the extension:
<img width="884" alt="image" src="https://user-images.githubusercontent.com/6349682/199210916-c089adf7-478a-4bda-8c98-9be8c8fa8fd3.png">

**Comment by mrk-han** [comment]
@mattjohnsonpint I'm really glad you started this issue, because I'm working through the same thing now. The pop-up is blocking a lot, and I'm not sure it's achieving VM acceleration.

The only difference I'm running into, is that I thought I was not getting a valid `accel-check` before running the command -- I will try to make a bare-bones repo and test this

**Reply by plaitik** [reply]
![20260129_040419-COLLAGE~2](https://github.com/user-attachments/assets/00c111ae-5582-4a91-80a9-73f7b5dceb0d)
,привет,решил ли ты проблему с этими окнами -подсказками?

**Comment by mrk-han** [comment]
I have been talking back and forth while trying to figure this stuff out with @igorboskovic3 who sent me https://github.com/igorboskovic3/actions-test/actions/runs/3228474831 which also has a system-extension updated popup. [Screenshot of Popup](https://github.com/igorboskovic3/actions-test/suites/8721062629/artifacts/394232381)

I really do think we need more clarity on this from the runner team.


Here are the few issues I have been communicating in related to HAXM:
[Update/Add HAXM to macOS 11 and macOS 12 agents](https://github.com/actions/runner-images/issues/6388)

[Android emulator boots up with black screen on microsoft hosted agent with mac-11 and Mac-12 in Azure pipeline](https://github.com/actions/runner-images/issues/6152)

[Getting spammed with [CAMetalLayer nextDrawable] returning nil because device is nil on latest macOS Agent](https://github.com/actions/runner-images/issues/6365)


Another thing I want to look into is potentially using the new SKIA Rendering `$ANDROID_HOME/platform-tools/adb shell 'su;setprop debug.hwui.renderer skiagl;stop;start'`  -- but it requires the emulator to be rooted I believe.

**Reply by tenzap** [reply]
I just tried locally on an image with gpu accel disabled:
```
adb shell "su root setprop debug.hwui.renderer skiagl"
adb shell "su root stop"
adb shell "su root start"
```
The UI didn't seem much more responsive than without `skiagl`

**Reply by mrk-han** [reply]
@tenzap This checks out. Locally our machines are much more powerful nowadays compared to CI, but the main difference is on CI we are running on VMs so it's nested virtualization with the emulator vm. Locally we obviously don't have that GPU passthrough constraint.

**Comment by mattjohnsonpint** [comment]
The main reason I am digging on this is that some of our Android tests are extremely flaky - but only when running in Github Actions.  Run the same tests on an emulator on either my Mac or Windows machines and they pass every time.  Even if I let it run repeatedly for hours.

Also sometimes the emulators in CI flat out fail to boot.

Also API 32 seems more flaky than others - not sure why.

Originally I thought perhaps the issue was because the warning we get:

```
 WARNING | Running on a system with less than 6 logical cores. Setting number of virtual cores to 1
```

But like I said, it always works locally - even on an AVD configured for a single core.

So just process of elimination - trying to figure out what is special about the CI environment that causes the flakiness.  Perhaps it's hardware acceleration, perhaps not.

**Reply by tenzap** [reply]
API32 & 33 are very flaky for me too. And there are no problems locally. Howerver if locally I disable gpu accel, I notice that the UI is way much slower.

**Reply by tenzap** [reply]
Maybe you are interested in this:

> "macOS emulator is super slow on machines with less than 6 logical cores."

https://android.googlesource.com/platform/external/qemu/+/refs/heads/emu-master-dev/android-qemu2-glue/main.cpp#2519
```
#ifdef __APPLE__
        // macOS emulator is super slow on machines with less
        // than 6 logical cores.
        if (System::get()->getCpuCoreCount() < 6) {
            dwarning(
                    "Running on a system with less than 6 logical cores. "
                    "Setting number of virtual cores to 1");
            hw->hw_cpu_ncore = 1;
        }
#endif
```

**Comment by mrk-han** [comment]
@mattjohnsonpint Same here. We have quite a few teams all trying to run affordable automation leveraging Github Actions + Github Action Runners after migrating from a self-hosted solution (KVM enabled Ubuntu Agent hosted on MacStadium)

I have also been super skeptical of why ` WARNING | Running on a system with less than 6 logical cores. Setting number of virtual cores to 1` is happening too.

And yes, I'm pretty sure I could run the test suite with 4x sharding locally on our M1/M2 MacBooks, so it's quite hilarious we're in this state on CI.

**Comment by mrk-han** [comment]
@mattjohnsonpint I'm starting to think there's much of an advantage at all with running on macOS agents. We used to run on them because HAXM was supported, and things were more "stable" on macOS than Ubuntu Agents (The standard Github ones)


If we're going to have the same performance, I'd rather run on 1/10th the cost Ubuntu agents than the macOS agents.

Also, the Ubuntu agents, if you have access to the larger agent beta, can provide way more powerful machines.

**Comment by mrk-han** [comment]
@mattjohnsonpint I just did a test and discovered that if you run `emulator -accel-check` with the new tools, e.g. `$ANDROID_HOME/emulator/emulator -accel-check` then it always passes, but if you run `emulator -accel check` with the old tools, then it is fails, e.g. `$ANDROID_HOME/tools/emulator`

This is more than confusing to me.

https://github.com/mrk-han/Emulator-Sandbox/actions/runs/3374542418

New Tools -> `"$ANDROID_HOME"/emulator/emulator -accel-check`

```
accel:
0
Hypervisor.Framework OS X Version 12.6
accel

```

Old Tools -> `"$ANDROID_HOME"/tools/emulator -accel-check`
```
accel:
6
HAXM is not installed on this machine
accel

```

**Comment by tenzap** [comment]
Maybe an alternative would be to use an Automated Test Device (ATD) image
https://developer.android.com/studio/releases/platforms#12

Only available currently on API30 & API31 (canary channel) though.
```
./cmdline-tools/latest/bin/sdkmanager --install "system-images;android-30;google_atd;x86"
./cmdline-tools/latest/bin/sdkmanager --channel=3 --install "system-images;android-31;google_atd;x86_64"
```

**Reply by ychescale9** [reply]
[this](https://github.com/android/nowinandroid/pull/479/files)
 shows how to run instrumented tests with Gradle managed device on GitHub actions.

**Reply by tenzap** [reply]
Thanks. I was mainly concerned here by the slowlyness & flakiness of images for recent apis

**Reply by ychescale9** [reply]
That's not something a tool like this can fix unfortunately.

**Reply by mrk-han** [reply]
Ahh finally the ATD images for Android 12 are out! Good to know. 

I did not have much success with the Android-30 ATD images on Github's macOS Runner.

**Comment by mrk-han** [comment]
See this comment: https://github.com/ReactiveCircus/android-emulator-runner/issues/46#issuecomment-1473089957
